validator/delphi_validator/validate.py

```diff
@@ -1,237 +1,721 @@
+                    ("check_se_missing_or_in_range", nameformat),
+                    result, "se must be NA or in (0, min(50,val*(1+eps))]"))
+
+        result_jeffreys = df_to_test.query('(val == 0) & (se == 0)')
+        result_alt = df_to_test.query('se == 0')
+
+        if not result_jeffreys.empty:
+            self.raised_errors.append(ValidationError(
+                ("check_se_0_when_val_0", nameformat),
+                None,
+                "when signal value is 0, se must be non-zero. please "
+                + "use Jeffreys correction to generate an appropriate se"))
+        elif not result_alt.empty:
+            self.raised_errors.append(ValidationError(
+                ("check_se_0", nameformat),
+                result_alt, "se must be non-zero"))
+
+    def check_bad_sample_size(self, df_to_test, nameformat):
+        """
+        Check sample sizes for validity.
+
+        Arguments:
+            - df_to_test: pandas dataframe of a single CSV of source data
+            (one day-signal-geo_type combo)
+            - nameformat: str CSV name; for example, "20200624_county_smoothed_nohh_cmnty_cli.csv"
+
+        Returns:
+            - None
+        """
+        if not self.missing_sample_size_allowed:
+            if df_to_test['sample_size'].isnull().values.any():
+                self.raised_errors.append(ValidationError(
+                    ("check_n_missing", nameformat),
+                    None, "sample_size must not be NA"))
+
+            # Find rows with sample size less than minimum allowed
+            result = df_to_test.query(
+                '(sample_size < @self.minimum_sample_size)')
+
+            if not result.empty:
+                self.raised_errors.append(ValidationError(
+                    ("check_n_gt_min", nameformat),
+                    result, "sample size must be >= {self.minimum_sample_size}"))
+
+        elif self.missing_sample_size_allowed:
+            result = df_to_test.query(
+                '~(sample_size.isnull() | (sample_size >= @self.minimum_sample_size))')
+
+            if not result.empty:
+                self.raised_errors.append(ValidationError(
+                    ("check_n_missing_or_gt_min", nameformat),
+                    result,
+                    "sample size must be NA or >= {self.minimum_sample_size}"))
+
+    def check_min_allowed_max_date(self, max_date, weighted_option, geo, sig):
+        """
+        Check if time since data was generated is reasonable or too long ago.
+
+        Arguments:
+            - max_date: date of most recent data to be validated; datetime format.
+            - weighted_option: str; selects the "reasonable" threshold based on signal name
+            - geo: str; geo type name (county, msa, hrr, state) as in the CSV name
+            - sig: str; signal name as in the CSV name
+
+        Returns:
+            - None
+        """
+        switcher = {
+            'unweighted': timedelta(days=1),
```

Comment:
Is this essentially the assumed lag? doctor-visits, hospital-admissions, and quidel will all have much longer minimum lags than this.

Reply:
My understanding is that there are two different types of "lag".

1. One is the number of days behind a data source is from being current. So, for example, a centered 7-day average will always be 3-4 days behind *now* because of how it's calculated.
2. The other type of lag, to be addressed with the new "scan_length" setting, happens when case (or other metric) reports are delayed compared to when they happen, causing recent already-released data to be updated.

 We're dealing with the first type here (going to be renamed to "lag_option" or similar), so I'm surprised that it varies widely. I thought the goal was to have all the indicators update on a daily basis and not be more than a few days out of date.

I'll plan to add this as a user setting so it can be set on a per-signal basis.

Reply:
Ah! Good note.

In the COVIDcast API, we define lag to be the number of time units between when an event occurred and when data about those events were published. It's a purely operational definition: `issue - time_value`.

Delphi seems to collectively define "backfill" as the effect where previously-reported values are updated as additional data about those events becomes available. Backfilled data by definition has greater lag than what was first reported.

These two terms do not exactly match the two definitions you're familiar with, but they have roughly the same coverage (with the possible exclusion of filter lag, but then the API hasn't had to explicitly handle that).

Reply:
> I thought the goal was to have all the indicators update on a daily basis and not be more than a few days out of date.

That was the goal in early April, but it quickly became clear that some sources wouldn't have data, or wouldn't have enough data, available quickly enough. iirc doctor-visits first publishes at 5 days out with 70% of the data it will ultimately have about that day; safegraph data has been up to 7 days out, etc.

Reply:
Perfect, thanks for the clarification!